File: leetcode/Unique_Paths.cpp
class Solution {
    public:

    int mm, nn;

    int solve(int m, int n, long long d[][100]) {
	    if (m == mm && n == nn) return d[m][n] = 1;
	    if (m > mm || n > nn) return 0;
	    if (d[m][n] != -1) return d[m][n];
	    ++d[m][n];
	    d[m][n] += solve(m+1, n, d);
	    d[m][n] += solve(m, n+1, d);
	    return d[m][n];
    }

    int uniquePaths(int m, int n) {
	    mm = m-1; nn = n-1;
	    long long d[100][100];
	    for (int i = 0; i < m; ++i)
	    for (int j = 0; j < n; ++j) d[i][j] = -1;
	    solve(0, 0, d);
	    return d[0][0];
    }
};

File: leetcode/UniquePaths.java

public class UniquePaths {

    private int rowLength;
    private int columnLength;
    private int d[][];

    public static void main(String[] args) {
	UniquePaths main = new UniquePaths();
	System.out.println(main.uniquePaths(100, 100));
    }

    public int uniquePaths(int m, int n) {
	if (m < 0 || n < 0) {
	    return 0;
	}
	if (m == 0 || n == 0) {
	    return 1;
	}
	rowLength = m;
	columnLength = n;
	d = new int[rowLength][columnLength];
	return finder(0, 0);
    }

    private int finder(int x, int y) {
	if (x >= rowLength || y >= columnLength) {
	    return 0;
	}
	if (x == rowLength - 1 && y == columnLength - 1) {
	    d[x][y] = 1;
	}
	if (d[x][y] > 0) {
	    return d[x][y];
	}
	d[x][y] = finder(x + 1, y) + finder(x, y + 1);
	return d[x][y];
    }
}
